### DISCUSSION TITLE
App Routerのディレクトリについてつらつら書く

### DISCUSSION THREAD

**Discussion by pss-aileen** [first-contributions-ja/first-contributions-ja.github.io#60] [90: 知識・知見の共有]
ちょっと情報整理をかねてこちらにつらつら書いてみます

---

- コンポーネントとして切り出せるものは切り出して componentsディレクトリに入れる
- page.tsx に書く、そのページ独自のテキストや繰り返さないものはコンポーネントにしなくても良い

というのが今時点の認識。
それをもとに考えるとディレクトリについての考え方が落ち着いてきそう

**Comment by pss-aileen** [comment]
サンプルとなるようなもの

- 調査データ探し
  - https://vercel.com/templates

**Reply by pss-aileen** [reply]
- [Next.js App Router Playground – Vercel](https://vercel.com/templates/next.js/app-directory)
  - Routerのあれこれが試せる？例を見れるやつ
  - まったく理解していない

**Reply by pss-aileen** [reply]
[leerob/leerob.io: ✨ My portfolio built with Next.js, Tailwind, and Vercel.](https://github.com/leerob/leerob.io)
- tailwindcssはそのまま書いている感じ
- フォントぐらいしかconfigに設定しない

**Comment by pss-aileen** [comment]
関係ないけど気になったもののリンク
- [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)
  - コミットメッセージの書き方、人間と機械が読みやすく、意味のあるコミットメッセージにするための仕様

**Comment by pss-aileen** [comment]
「コンポーネント」とは何か、コンポーネントとして書き出すものは何が適切か？

**Reply by pss-aileen** [reply]
- page.tsxに書いても良いもの（別に切り分けなくても良いもの）
  - TOPページ特有のコンポーネント
    - ヒーローエリア、ABOUTセクション
    - TOPページがもっと長くなって、セクションを使い回す場合はコンポーネント化した方が良い？という認識
      - [Blazity/next-enterprise](https://github.com/Blazity/next-enterprise)
        - [app/page.tsx](https://github.com/Blazity/next-enterprise/blob/main/app/page.tsx) にそのページだけで表示させる内容はそのまま書いている 
-  layout.tsx から切り離してコンポーネント化するもの
  - ヘッダー
  - フッター
  - 今は量が少ないので↑を切り離さなくても良いかな、とも思うけど
  - これらは切り離した方が良いと思っている。ただ、具体的に理由を説明できない

**Reply by pss-aileen** [reply]
そもそも
- layout.tsx
  - 全てのディレクトリに継承できるレイアウト
- page.tsx
  - TOPページの内容（各ページで表示される内容を記述する）
  - HTMLでいう、index.htmlの内容
  - コンポーネント化するというのは、HTMLの中でも使い回すものをコンポーネント化して再利用しやすくするためにある、はず？
  - なので、コンポーネントとして切り分けるのは「再利用するもの」といった認識がよいのか

**Reply by pss-aileen** [reply]
- [steven-tey/precedent: An opinionated collection of components, hooks, and utilities for your Next.js project.](https://github.com/steven-tey/precedent)
  - これも app/page.tsx にTOPの内容がそのまま書き込んである
  - componentsは繰り返して使用したいものが使われている模様
  - layoutのフォルダ構成参考になるかも
  - <img src="https://github.com/first-contributions-ja/first-contributions-ja.github.io/assets/140990100/efb87cb1-c570-49b5-886f-cd52882d19d4" width="200">

**Comment by kazzyfrog** [comment]
自分もこの前App Routerのディレクトリ構成色々調べたので、後でつらつら書きます！

**Reply by kazzyfrog** [reply]
以前書いた分かりずらいメモだけ、一旦貼り付けておきますw

## ディレクトリ構成

[Shadcn/ui](https://ui.shadcn.com/docs/installation/next#app-structure)に合わせて、[プロジェクトファイルを`app`の外部に保存](https://nextjs.org/docs/app/building-your-application/routing/colocation#store-project-files-outside-of-app)するパターン

そして、`app`ディレクトリ内は3つの階層に分かれています：

1. [Next.jsのメタデータファイル](https://nextjs.org/docs/getting-started/project-structure#routing-files)
2. 基本的なサイトのページ構造
3. [コロケーションの概念](https://kentcdodds.com/blog/colocation)

以下のようになります：
```
    1   2   3

├─app
|   ├─(pages)
|   |   ├─(root)
|   |   │   ├─page.tsx
|   |   │   ├─_components
|   |   │   │   └─componentA.tsx
|   |   │   └─_types
|   |   |       └─types.ts
|   |   ├─pageA
|   |   |   ├─page.tsx
|   |   |   ├─_components
|   |   |   │   └─componentB.tsx
|   |   |   └─_types
|   |   |      └─types.ts
|   |   └─layout.tsx
|   ├─error.tsx
|   ├─layout.tsx
|   ├─loading.tsx
|   └─not-found.tsx
|
├─components
|   ├─navbar.tsx
|   └─ui
|     └─button.tsx
├─libs
|
```

**Reply by kazzyfrog** [reply]
ページ数が少ない・規模が小さい場合は、`app`と同じ階層に`components `フォルダを用意するだけで良さそうです！
各ページディレクトリの中の`_components`は、複雑になってきてからで！

**Comment by pss-aileen** [comment]
## Reactとは

### コンポーネントを作成、ページで組み合わせられる

> React lets you build user interfaces out of individual pieces called components. Create your own React components like Thumbnail, LikeButton, and Video. Then combine them into entire screens, pages, and apps.

訳: Reactは、個々の部品を「コンポーネント」と呼ばれるもので構築することで、ユーザーインターフェースを構築することができます。Thumbnail、LikeButton、Videoのような独自のReactコンポーネントを作成し、それらを画面、ページ、アプリ全体に組み合わせることができます。

### 画面上に表示される要素を作るためライブラリ、UIコンポーネントを作るためのツール

> React is a JavaScript library created by Facebook
> React is a User Interface (UI) library
> React is a tool for building UI components

訳:
- ReactはFacebookによって作られたJavaScriptライブラリ
- Reactは、UI（ユーザーインターフェース）ライブラリ
- Reactは、UIコンポーネントを作るためのツール

#### 意味
- UI とは
  - User Interface
  - 画面上に表示されるボタンやメニュー、フォーム、ユーザーが操作する部分全体
 
## Next.jsとは
GPTより
> Next.jsは、Reactアプリケーションのフレームワークであり、ページベースのルーティングやサーバーサイドレンダリングなどの機能を提供します。Next.jsを使用すると、Reactコンポーネントを使ってページのUIを作成し、それをページコンポーネントとしてエクスポートすることができます。
